pep-0639.rst

```diff
@@ -1,543 +1,2209 @@
+reject any that specify a ``License`` field and the text of which is not
+identical to that of ``License-Expression``, as defined in this section.
 
-- SHOULD report an informational warning if normalization process results in
-  changes to the ``License`` field contents.
+Along with license classifiers, the ``License`` field may be removed from a
+new version of the specification in a future PEP.
 
-License expression examples::
 
-    License: MIT
+Deprecate license classifiers
+'''''''''''''''''''''''''''''
 
-    License: BSD-3-Clause
+Using license `classifiers <#classifiers_>`_ in the ``Classifier`` field
+(described in PEP 301) is deprecated and replaced by the more precise
+``License-Expression`` field.
 
-    License: MIT OR GPL-2.0-or-later OR (FSFUL AND BSD-2-Clause)
+If the ``License-Expression`` field is present, build tools SHOULD and
+publishing tools MUST raise an error if one or more license classifiers
+is included in a ``Classifier`` field, and MUST NOT add
+such classifiers themselves.
 
-    License: GPL-3.0-only WITH Classpath-Exception-2.0 OR BSD-3-Clause
+Otherwise, if this field contains a license classifier, build tools MAY
+and publishing tools SHOULD issue a warning informing users such classifiers
+are deprecated, and recommending ``License-Expression`` instead.
+For compatibility with existing publishing and installation processes,
+the presence of license classifiers SHOULD NOT raise an error unless
+``License-Expression`` is also provided.
 
-    License: This software may only be obtained by sending the
-            author a postcard, and then the user promises not
-            to redistribute it.
+For all newly-uploaded distributions that include a
+``License-Expression`` field, the `Python Package Index (PyPI) <#pypi_>`_ MUST
+reject any that also specify any license classifiers.
 
-    License: LicenseRef-Proprietary AND LicenseRef-Public-Domain
+New license classifiers MUST NOT be `added to PyPI <#classifiersrepo_>`_;
+users needing them SHOULD use the ``License-Expression`` field instead.
+Along with the ``License`` field, license classifiers may be removed from a
+new version of the specification in a future PEP.
 
 
-Classifier (multiple use)
-:::::::::::::::::::::::::
+Project source metadata
+-----------------------
 
-Each entry is a string giving a single classification value for the
-distribution. Classifiers are described in PEP 301.
+As originally introduced in PEP 621, the
+`PyPA Declaring Project Metadata specification <#pep621spec_>`_
+defines how to declare a project's source
+metadata in a ``[project]`` table in the ``pyproject.toml`` file for
+build tools to consume and output distribution core metadata.
 
-Examples::
+This PEP `adds <Add license-expression key_>`_ the ``license-expression`` key,
+`adds <Add license-files key_>`_ the ``license-files`` key and
+`deprecates <Deprecate license key_>`_ the ``license`` key.
```

Comment:
PEP 621 specifically left room in the `license` key, so we don’t need to deprecate it. We only need to deprecate `license = {text = ...}` and `license = {file = ...}`, and define meaning to `license = "MIT"`. No new `[project]` keys are needed.

To allow multiple licenses, we add

```toml
license = { names = ... }
```

Also, I think we should always glob to avoid confusion. So we define

```toml
license = { files = ... }
```

where each string is a glob pattern, and if someone really really need stars in the path (extremely unlikely), they can escape them manually.

Reply:
> PEP 621 specifically left room in the license key, so we don’t need to deprecate it. We only need to deprecate license = {text = ...} and license = {file = ...}, and define meaning to license = "MIT". No new [project] keys are needed.

This was exactly the original approach I took here (see 4871854 ), later revised to split the `license-files` out into a top-level key while keeping the `License-Expression` field mapped to the `license` key, and further revised to split them both (see  88cba0c ). In practice, when considered more carefully, the first, and to a somewhat lesser extent the second approach provided to be distinctly suboptimal for a number of reasons, and quite possibly unworkable (due to the interaction with the `dynamic` key, depending on exactly how one interprets a rather imprecisely specified passage in PEP 621), while the only downside of the present approach was 1-2 extra top-level keys.

I spent a substantial amount of time carefully considering the alternatives here, and thoroughly address both of these in great detail in the `Rejected Ideas` section; see the `Add expression and files subkeys to table` and `Define license expression as string value` items under the `Source metadata license key` subsection. If you still have articulatable concerns with this approach, or can find ways around the various issues I've cited with the previous one, it might be better to discuss this on the Discourse thread, where we can all come to a clear consensus before I spend several more hours of my free time changing everything back (and us all having to bikeshed over exactly what to change it back _to_, heh).

> To allow multiple licenses, we add `license = { names = ... }`

Perhaps I'm grossly misunderstanding your suggestion, but this proposal appears to contradict the underlying goal and basic objective of this PEP—adding SPDX license expressions to Python package metadata—which has been true its conception, has not been (at least as far as I could find) brought up throughout the course of the years-long relevant discussions, and is intrinsically incompatible with the core metadata field `License-Expression` that it is supposed to map to. Could you explain a little more about what you're suggesting here? Given it seems to be a fairly substantial shift in what this PEP has been proposing to do from the start, and which seemed to have a fairly strong consensus behind it, if you're really sure about it, you might want to bring it up on Discourse first.

> Also, I think we should always glob to avoid confusion.

I think I'm the one that's a little confused, heh—could you explain a little more how users having to implicitly know that `license.files` is a glob is _less_ potentially confusing than them explicitly specifying ``license-files.glob`` (or ``license-files.paths``)? Furthermore, per the spec, for the latter tools will unambiguously know if the user mistakenly specify a glob instead of a literal path under the `paths` key, because it won't match an actual file.

In any case, like the previous, this was the approach I initially implemented here but after a substantial amount of careful thought and experimentation, this ran into significant conceptual and practical issues with, for a few extra characters, the more explicit alternative neatly avoids. I address this issue at length under the `Only accept glob patterns` and `Infer whether paths or globs` items in the `Source metadata license-files key` subsection of the `Rejected Ideas` section.

> if someone really really need stars in the path (extremely unlikely), they can escape them manually.

Actually, this is not the main motivation for this choice, or even one of the primary drivers. The `Only accept glob patterns` and `Infer whether paths or globs` items in the `Source metadata license-files key` subsection of the `Rejected Ideas` section describes the rationale in detail.

Reply:
> Perhaps I'm grossly misunderstanding your suggestion, but this proposal appears to contradict the underlying goal and basic objective of this PEP—adding SPDX license expressions to Python package metadata—which has been true its conception, has not been (at least as far as I could find) brought up throughout the course of the years-long relevant discussions, and is intrinsically incompatible with the core metadata field `License-Expression` that it is supposed to map to.

No, you’re right, I was misunderstanding how `License-Expression` works (I thought it could be specified multiple times to automatically mean “pick one of these”).

> I think I'm the one that's a little confused, heh—could you explain a little more how users having to implicitly know that `license.files` is a glob is less potentially confusing than them explicitly specifying `license-files.glob` (or `license-files.paths`)?

It’s not, but there’s no harm for that :) It is extremely uncommon for license files to be put in paths that resemble a glob pattern, so the user does not actually care if this triggers globbing when they only want to specify one file. And if they want globbing, now they have only one key to remember instead of two.

---

Perhaps this would be a better approach?

1. Add `license-files` to do glob (no nested keys).
2. Use `license = "<spdx-expression>"` and deprecate nested keys.

Reply:
> No, you’re right,

No worries, sorry for the confusion. Seems like the mixup might have been over a license identifier vs. a license expression; I do define these in the [Terminology](https://github.com/python/peps/blob/6aab97e06c1ca81b14a3d57d44cea562c64aad38/pep-0639.rst#terminology) section, but is there anything in the PEP that I should clarify here?

> It is extremely uncommon for license files to be put in paths that resemble a glob pattern, so the user does not actually care if this triggers globbing when they only want to specify one file.

Indeed; however, as discussed this isn't the (or even a) primary reason for this; sorry for not linking it when I referred to it previously, but I discuss the rationale in detail in the [Only accept glob patterns](https://github.com/python/peps/blob/6aab97e06c1ca81b14a3d57d44cea562c64aad38/pep-0639.rst#only-accept-glob-patterns) and [Infer whether paths or globs](https://github.com/python/peps/blob/6aab97e06c1ca81b14a3d57d44cea562c64aad38/pep-0639.rst#infer-whether-paths-or-globs) items in the Rejected Ideas, and see next item for a summary.

> Add license-files to do glob (no nested keys).

Actually, as mentioned, this was also my initial approach, as "flat is better than nested", but it ended up conflicting with several of the behaviors specified so that "errors should never pass silently", especially if "in the face of ambiguity we must resist the temptation to guess", and in the end, I decided "explicit is better than implicit" and that the structure being slightly more "complex is better than complicated" logic to parse it :) See 

> Use `license = "<spdx-expression>"` and deprecate nested keys.

In fact, following that which you previously proposed, this was my second approach here too :) However, upon implementing it, it still ran into less serious versions of some of the same issues as the previous, including mapping two distinct metadata types to the same key, inability to specify which was marked as dynamic, and creating additional inconsistency between core metadata, PEP 621 and ``setup.cfg``/other tool formats. On top of that, it creates potential for serious user confusion and ambiguity, since it means ``license`` maps to completely syntactically and semantically different fields depending on the config format, and is much less explicit to authors, readers and consumers that it is intended to be a SPDX identifier/expression. See [Define license expression as string value](https://github.com/python/peps/blob/6aab97e06c1ca81b14a3d57d44cea562c64aad38/pep-0639.rst#define-license-expression-as-string-value) in the rejected ideas for more details.